.github/gha_monitor/gha_monitor/__main__.py

```diff
@@ -0,0 +1,109 @@
+
+
+class GitHubActions(github.GitHubClient):
+    params = {
+        # Needed when using an API key
+        'github_username': os.getenv('github_username', None),
+        'github_password': os.getenv('github_password', None),
+        # Use from within an Action - ignored if username set
+        'token': os.getenv('GITHUB_TOKEN', None),
+        'repo_organization': os.getenv('GITHUB_REPO_ORG'),
+        'repo': os.getenv('GITHUB_REPO')
+    }
+
+
+class S2nNotices(sns.SNSClient):
+    params = {
+        'topic_arn': 'arn:aws:sns:us-west-2:024603541914:s2n_notices'
+    }
+
+
+def message_text():
+    """ Formatting for text message. """
+    return """
+s2n GitHub Action monitor notice
+State: {conclusion}
+Repo: {repo}
+GHA failure time: {time}
+Workflow name: {workflow_name}
+URL: {url}
+started by: {commit_owner}\n
+"""
+
+
+def main():
+    """ Main entrypoint. """
+    logging.info('Starting up')
+    plaintext_notice = []
+    gh_api = GitHubActions()
+    s2n_text_client = S2nNotices()
+
+    # Get the Action workflow log from the Github API
+    gh_api.get_workflow_log_chunk(final_state='failure')
+    logging.info(f"Looking for failures newer than {TIME_WINDOW_BEGIN} in {gh_api.params['repo_organization']}"
+                 f"/{gh_api.params['repo']}")
+    if gh_api.worklog:
+        for enhanced_worklog in gh_api.worklog:
+            # Parse the event date/time so we can compare it
+            datetime_creation = parser.parse(enhanced_worklog['created_at'])
+            logging.debug(f"looking at event from {enhanced_worklog['created_at']}")
+
+            # If the event is recent enough, process it.
+            if datetime_creation > TIME_WINDOW_BEGIN:
+                logging.debug(f"Workflow_url: {enhanced_worklog['workflow_url']}")
+                # The name of the workflow isn't in the failure object, look it up.
+                enhanced_worklog['workflow_name'] = gh_api.get_workflow_name(enhanced_worklog['workflow_url'].split('/')[-1:][0])
+                enhanced_worklog['repo'] = gh_api.params['repo']
+
+                # Construct a notification string.
+                notice_msg = message_text().format(
+                    conclusion=enhanced_worklog['conclusion'],
+                    time=enhanced_worklog['created_at'],
+                    url=enhanced_worklog['html_url'],
+                    commit_owner=enhanced_worklog['head_commit']['author']['email'],
+                    repo=enhanced_worklog['repo'],
+                    workflow_name=enhanced_worklog['workflow_name'])
+                logging.debug(notice_msg)
+                plaintext_notice.append(notice_msg)
+            else:
+                logging.debug("event outside time range.")
+    else:
+        logging.info("GH API returned empty worklog")
+
+
+    # Relay messages to SNS
+    if plaintext_notice:
+        # Combine multiple message together.
+        logging.info(s2n_text_client.publish("\n".join(plaintext_notice)))
```

Comment:
Should these be collapsed into a single message? It seems to me like it would be more manageable to have each sent as a separate message, both for handling these events on the receiving end and for scaling. Throwing some guesstimate data into the template, I'm getting a single entry as ~250-300 bytes. You might not exceed ~800 failed events in a single run, but do you want to bank on that? And on the receiving end, do you really want to have to parse each event out of the larger combined message?